Add cmocean color maps for E3SM team

markcmiller86 opened the pull request:
### Description

Resolves #20403

<!-- Please include a summary of the change and any other information and instructions to help reviewers understand the work -->

<!-- Note that all the checklist items in various bulleted lists below end with ~~ characters.
     This is to facilitate striking them out when they do not apply.
     One just has to add ~~ characters just before the opening square bracket [ -->

### Type of change

<!-- Please check one of the boxes below -->

* ~~[ ] Bug fix~~
* [x] New feature~~
* ~~[ ] Documentation update~~
* ~~[ ] Other~~ <!-- please explain with a note below -->

### How Has This Been Tested?

<!-- Please describe the tests you've added or any tests that already cover this change. Include relevant information, such as which operating system you tested on. -->

### Reminders:

- Please follow the [style guidelines][1] of this project.
- Please perform a self-review of your code before submitting a PR and asking others to review it.
- Please assign reviewers (see [VisIt's PR procedures][2] for more information).

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- ~~[ ] I have commented my code where applicable.~~
- [x] I have updated the release notes.~~
- ~~[ ] I have made corresponding changes to the documentation.~~
- ~~[ ] I have added debugging support to my changes.~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works.~~
- ~~[ ] I have confirmed new and existing unit tests pass locally with my changes.~~
- ~~[ ] I have added new baselines for any new tests to the repo.~~
- [x] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

markcmiller86 committed:
add cmocean color maps

markcmiller86 committed:
add note

JustinPrivitera commented on data/vtk_test_data.tar.xz:
What is this test data for?

markcmiller86 replied:
Whoops! Hmmmm...I think I know what happened. That is to update some test data to test files with unicode characters in variable names and path names. I forgot I had that uncommitted on develop when I created this branch.

JustinPrivitera replied:
So adding this tarball is supposed to be part of this merge? Just checking.

markcmiller86 replied:
Sorry...working on backing this one file out.

markcmiller86 replied:
Ok, fixed this now.

JustinPrivitera commented on src/resources/colortables/algae.ct:
Did you verify that these color tables appear as you expect in VisIt? I don't think the "cmocean" tag is sufficient. To match the other tags it ought to be capitalized: "Cmocean". We should also give each one either the "Continuous" or "Discrete" tag.

markcmiller86 replied:
Yes, I did make sure they load up correctly. Though...now that I think about it, we should have some standard object plots that we render in a given color table to compare to some standard image obtained from using the table in its original form too. That would be a good test. I will add a ticket for that...to test our color tables.

The name `cmocean` is not itself capitalized and so I didn't want to capitalize it. I don't think its necessary to capitalize tags, is it?

These are all continuous. Why a tag for this when `discrete=true` present as a separate XML field value already identifies whether the table is discrete or, if not present, continuous?

JustinPrivitera replied:
> The name `cmocean` is not itself capitalized and so I didn't want to capitalize it. I don't think its necessary to capitalize tags, is it?

No, but it'd be the only tag that is not capitalized. Why don't we call it "CMOcean".

> These are all continuous. Why a tag for this when `discrete=true` present as a separate XML field value already identifies whether the table is discrete or, if not present, continuous?

I didn't see the `discrete=true` XML field value in all your color table files, but maybe I missed it. Not all of our color tables have that field, as far as I know. For now, we have been putting `Discrete` or `Continuous` into each color table tags field. We can open a ticket if we want to change that practice and have VisIt automatically figure it out from the XML field. I am happy to investigate that, but let's make things similar to the other color tables for now.

markcmiller86 replied:
> No, but it'd be the only tag that is not capitalized.

I think I am ok with that.

> Why don't we call it "CMOcean".

Must we? My preference is to spell a product's moniker the way the product owner's do. I think it makes it more recognizeable to the users who care about it.
 
> I didn't see the `discrete=true` XML field value in all your color table files

Correct. You shouldn't. That is because these are all continuous color tables.

> Not all of our color tables have that field, as far as I know.

Correct. The ones that are indeed discrete do. Otherwise, they are continuous in which case, the line is not present (to be clear the *continous* case is **not** handled by having a line indicating `discrete=false`)...not sure why but that is my read based on the files already there.

> For now, we have been putting `Discrete` or `Continuous` into each color table tags field.

I think we should avoid duplicating information in tags that is already known elsewhere. Is the presence of `Discrete` or `Continuous` in tags used to prime the GUI's `discrete` and `continuous` check boxes? I think the XML tag's presence (or absence) is. What happens if someone adds both `Discrete` *and* `Continous` tags to a color table. Which is it then? Thinking about this...our GUI for this aspect of a color table is a tad off. Its got what appear to be independent check boxes for each but they are not *independent*. Its one or the other...like a multiple choice box (or pull down).

> We can open a ticket if we want to change that practice and have VisIt automatically figure it out from the XML field.

So, I just tested whether having `Discrete` or `Continous` in tags effects the GUI's selection of the `Color table type` check boxes. They don't. I believe that is driven by the XML field's presence or absence. I am not even sure cmocean color table users would even know what `Discrete` or `Continous` key words in tags for their color maps would actually mean. That feels like the VisIt project imposing a *requirement* on the use of tags when I thought the whole intention was for these to be pretty much arbitrary ways of classifying color tables in ways that make sense to individual users or user groups.

I also noticed another subtle issue...a color table's name in the GUI is taken directly from its file name...not XML file content. But, we also currently put all files into *one* directory. That means two communities could easily collide on color table names and one would be forced to choose a different name. This would be easily solved if we could put color tables in sub-directories. Then the directory name could serve as a kind of tag. A good choice might be to name the directories with the source of the color tables...`default`, `Cremari`, `cmocean`, `viridis` 

I can put these observatiosn into a new ticket if desired.

JustinPrivitera replied:
My argument is the same argument that you usually make for me: we have a way of doing things right now that is well-established. If we want to change that way of doing things, we should discuss and make changes across the board to how color tables work and how they are organized. Having a specific way of doing things for _just these color tables_ is not a good idea. When we add things we should make them look like everything else similar to them.

> > No, but it'd be the only tag that is not capitalized.
> 
> I think I am ok with that.
> 
> > Why don't we call it "CMOcean".
> 
> Must we? My preference is to spell a product's moniker the way the product owner's do. I think it makes it more recognizeable to the users who care about it.

I think providing a single tag that doesn't start with a capital letter looks sloppy. Users can add tags that don't start with capital letters, but all of ours that we provide are that way. I also think users will know that when we say "Cmocean" or "CMOcean" (which I'm guessing stands for "Color Maps Ocean"), that we mean "cmocean".

> I think we should avoid duplicating information in tags that is already known elsewhere. Is the presence of `Discrete` or `Continuous` in tags used to prime the GUI's `discrete` and `continuous` check boxes? I think the XML tag's presence (or absence) is.

I agree we should avoid duplicating information, but if I recall correctly this whole system is complicated, and automatically adding tags in the right cases has some tricky complexities. It is very possible that there is a better way to do it that I didn't think about when I was a brand new employee working on this.

I have a question for you: All of our color tables are either Continuous or Discrete. Users expect that if they select the Continuous tag they will see the Continuous color tables, and if they select the Discrete tag they will see the Discrete color tables. If you have a color table that has neither of those two tags in the XML, then does it appear in the color table manager when you ask to see only the Continuous tables? If it does appear there, then that means that I did my job well as a new employee and went the extra mile and came up with a way to automatically add that tag in if it is missing. If it does not appear there, then I would argue that your color table XML file is improperly specified. We can look to change the way VisIt populates the tags for each color table, but we should do that uniformly and change all color tables together instead of just having this set of tables be different.

> What happens if someone adds both `Discrete` _and_ `Continous` tags to a color table. Which is it then?

My guess is that an error is thrown, but I don't recall where that happens in the code.

> Thinking about this...our GUI for this aspect of a color table is a tad off. Its got what appear to be independent check boxes for each but they are not _independent_. Its one or the other...like a multiple choice box (or pull down).

I don't really understand. What part of the GUI are you referring to? The tag manager doesn't know that a table can't be both or neither, but other parts of the color table manager know.

> So, I just tested whether having `Discrete` or `Continous` in tags effects the GUI's selection of the `Color table type` check boxes. They don't. I believe that is driven by the XML field's presence or absence.

That's good; that makes sense.

> I am not even sure cmocean color table users would even know what `Discrete` or `Continous` key words in tags for their color maps would actually mean. That feels like the VisIt project imposing a _requirement_ on the use of tags when I thought the whole intention was for these to be pretty much arbitrary ways of classifying color tables in ways that make sense to individual users or user groups.

The cmocean users may not know what Continuous or Discrete mean, but they don't have to. But like I said before, if the color tables are indeed Continuous, and a user wants a Continuous color table, if the cmocean tables do not show up in the list of Continuous tables, then that is a bug. The way that VisIt reads color table files and populates the tags now is not necessarily wrong but could be changed. The bug is that the XML file is insufficiently specified for how VisIt currently expects to read it.

We do impose some (very minor) requirements on color tables so that they fit into our ecosystem. If we want to change that, let's open a ticket to discuss and figure out the best path forward.

> I also noticed another subtle issue...a color table's name in the GUI is taken directly from its file name...not XML file content. But, we also currently put all files into _one_ directory. That means two communities could easily collide on color table names and one would be forced to choose a different name. This would be easily solved if we could put color tables in sub-directories. Then the directory name could serve as a kind of tag. A good choice might be to name the directories with the source of the color tables...`default`, `Cremari`, `cmocean`, `viridis`

I think that is definitely worth changing.

> I can put these observatiosn into a new ticket if desired.

I think this is a good idea. I suggest we make these new color tables match how the others are for now, and if there are changes we want to make we can make them to all tables across the board as well as to how VisIt reads tables.

JustinPrivitera replied:
I glanced at [their paper](http://tos.org/oceanography/assets/docs/29-3_thyng.pdf) and it appears these color tables ought to have other tags as well:
![image](https://github.com/user-attachments/assets/ac3cc3a5-9094-4b03-ad57-89445a01574a)

![image](https://github.com/user-attachments/assets/89b7f3de-89a7-4f33-8210-4afa802b51a2)
![image](https://github.com/user-attachments/assets/b40a2400-7b83-406b-bdbf-da072944490f)

It is not immediately clear to me which ones fall into which category.

markcmiller86 replied:
Thanks for looking at the actual paper. I hand't bothered. Having looked now, I'll paste what I think is the most relevant paragraph in terms of understanding the attributes (and therefore useful tags) associated with this suite of colormaps...

> Figure 2 shows the cmocean collection, which is composed of several sequential colormaps meant to
elicit intuitive understanding of common oceanographic variables; three divergent colormaps; one cyclic colormap; and one hybrid colormap designed for the special case of displaying oxygen saturation. The package combines original colormaps developed specifically for this work with several preexisting colormaps (Moreland, 2009; Niccoli, 2012; Brewer, 2013; Samsel et al., 2015; see Acknowledgments); we have altered these maps for perceptual uniformity using viscm. Single-hue and multi-hue colormaps are included, and each of the sequential and diverg- ing colormaps span a wide range of lightness to maximize dynamic range in data display. The cmocean colormaps have been given names such as thermal, haline, and ice to help guide users to intuitive colormaps for common oceanographic variables; however, our nomenclature is not intended to restrict usage to any particular variable.

From this, I concluded we need to use tags, `Sequential`, `Divergent`, `Cyclic` and maybe `Singlue-hue` and/or `Multi-hue` (unless those are synonyms for tags we're already using).

Also, now that I notice the verbiage here... *several preexisting colormaps (Moreland,...Semsel)*, I worry we may already have some of these maps in VisIt by different names.

markcmiller86 replied:
> Also, now that I notice the verbiage here... _several preexisting colormaps (Moreland,...Semsel)_, I worry we may already have some of these maps in VisIt by different names.

Which suggests we need an additional feature...aliases for names.

markcmiller86 commented:
It would be good to put color tables in directories in the resources/colortables dir instead of all in one dir.

vijaysm commented:
Thank you so much @markcmiller86    👍

markcmiller86 committed:
adj tags

markcmiller86 committed:
fix spelling

markcmiller86 commented:
@JustinPrivitera ok I have updated requested...

* `cmocean` --> `CMOcean`
* Added `Continuous` tag to all
* Added `Sequential` tag to most
* Added `Diverging` tag to a few
* Added `Cyclic` tag to one

I used ChatGPT to tell me which color maps had which property.

markcmiller86 commented:
Oh, I also added `Color Blind Safe` tag to all of these too.

JustinPrivitera commented:
> @JustinPrivitera ok I have updated requested...
> 
>     * `cmocean` --> `CMOcean`
> 
>     * Added `Continuous` tag to all
> 
>     * Added `Sequential` tag to most
> 
>     * Added `Diverging` tag to a few
> 
>     * Added `Cyclic` tag to one
> 
> 
> I used ChatGPT to tell me which color maps had which property.

Thanks for this. ChatGPT's guess is as good as yours or mine.

> several sequential colormaps meant to
elicit intuitive understanding of common oceanographic variables; three divergent colormaps; one cyclic colormap; and one hybrid colormap

Perhaps we also need to add a "Hybrid" tag for whichever table is hybrid? It is as simple as including that tag name in the XML file.

> Oh, I also added `Color Blind Safe` tag to all of these too.

I don't remember seeing that in the paper but I did not read closely. Good to add that tag.

JustinPrivitera commented on src/resources/colortables/balance.ct:
```diff
@@ -1026,6 +1026,6 @@
            <Field name="position" type="float">1.000000</Field>
        </Object>
-    <Field name="tags" type="stringVector">"cmocean" </Field>
+    <Field name="tags" type="stringVector">"CMOcean" "Continuous" "Color Blind Safe" "Sequential" "Divergin"</Field>
```
"Divergin" should be "Diverging"

JustinPrivitera commented on src/resources/colortables/curl.ct:
```diff
@@ -1026,6 +1026,6 @@
            <Field name="position" type="float">1.000000</Field>
        </Object>
-    <Field name="tags" type="stringVector">"cmocean" </Field>
+    <Field name="tags" type="stringVector">"CMOcean" "Continuous" "Color Blind Safe" "Sequential" "Divergin"</Field>
```
Same here

JustinPrivitera commented on src/resources/colortables/delta.ct:
```diff
@@ -1026,6 +1026,6 @@
            <Field name="position" type="float">1.000000</Field>
        </Object>
-    <Field name="tags" type="stringVector">"cmocean" </Field>
+    <Field name="tags" type="stringVector">"CMOcean" "Continuous" "Color Blind Safe" "Sequential" "Divergin"</Field>
```
same here

JustinPrivitera reviewed (changes requested):
Looks good aside from the typo

JustinPrivitera commented:
Thanks for these changes and for discussing everything with me. Might have been better to talk in-person or over webex but I think we got to a good place. Looking forward to working through the ideas you brought up in your new ticket.

markcmiller86 commented:
Sorry. Saw those. Fixed 'em. Committed. But, didn't push.

markcmiller86 committed:
Merge branch 'develop' into feat-mcm86-21may25-cmocean-colormaps

markcmiller86 commented:
My understanding from the paper was that all the color maps were designed to handle moderate CVD (50%).

markcmiller86 committed:
revert changes to this file

markcmiller86 committed:
Merge branch 'feat-mcm86-21may25-cmocean-colormaps' of github.com:visit-dav/visit into feat-mcm86-21may25-cmocean-colormaps

JustinPrivitera commented:
Last thing - did we want to add a "Hybrid" tag for which ever color table is supposedly "hybrid"? Otherwise all looks good.

markcmiller86 commented:
> Last thing - did we want to add a "Hybrid" tag for which ever color table is supposedly "hybrid"? Otherwise all looks good.

I don't know what Hybrid means. Worse, because hybrid is pretty generic (it basically means a mixture...but a mixture of what) I worry it could mean one thing for cmocean color tables and entirely different thing for another user base.

But, I didn't look closely enough at meaning to know for sure if its fairly well pinned down or somewhat arbitrary.

JustinPrivitera commented:
> > Last thing - did we want to add a "Hybrid" tag for which ever color table is supposedly "hybrid"? Otherwise all looks good.
> 
> I don't know what Hybrid means. Worse, because hybrid is pretty generic (it basically means a mixture...but a mixture of what) I worry it could mean one thing for cmocean color tables and entirely different thing for another user base.
> 
> But, I didn't look closely enough at meaning to know for sure if its fairly well pinned down or somewhat arbitrary.

Fair enough. I don't know what it means either and you bring up a good point about it meaning different things to different people. I just saw it in the paper alongside the names of tags that we do have.

markcmiller86 commented:
So, ChatGPT is telling me that Hybrid in cmocean colormaps (which applies to `oxy` map only I think) is that it combines features of both `Sequential` and `Diverging`. So, I think simply including both tags on that map achieves the notion of `Hybrid`. Thoughts?

JustinPrivitera reviewed (dismissed):
Thanks for the iteration and discussion on this. Looks good. 👍

JustinPrivitera commented:
> So, ChatGPT is telling me that Hybrid in cmocean colormaps (which applies to `oxy` map only I think) is that it combines features of both `Sequential` and `Diverging`. So, I think simply including both tags on that map achieves the notion of `Hybrid`. Thoughts?

ChatGPT has never been wrong before. 🙃 That sounds reasonable so let's add both tags.

markcmiller86 commented:
Ok, lemme check I've got `Sequential` tag set correctly on those  that need it and not on those that don't

markcmiller86 committed:
fix tags
